Add ./check to print out useful current versions

jurraca opened the pull request:
prints Python, rpki-client, Kartograf, and Nix versions.

Unlike the issue, this adds a `./check` at the root of the project instead of `./run check`, seemed to be a little cleaner.

closes #18

fjahr commented:
The way I envisoned it, it would not just print the versions but also tell you if you have everything set up alright. I.e. if it's not the right version tell the users to which versions they should upgrade.

I think this could also check if all python packages are installed?

jurraca commented:
went with a different approach. Seemed reasonable to make it a test, but could just make it script.
problems:
- not sure how get the local RPKI version from CLI. Tried to capture output from a subprocess running "rpki-client -V" but couldn't figure it out.
- annoying little period on test output is messing with the alignment.

fjahr commented:
Test seems good for me! I want to have a real test suite eventually so might as well make the first steps.

> not sure how get the local RPKI version from CLI.

I implemented it like this for the log outputs we generate on the Version, I don't think anyone complained so far that it wasn't working so it should be reasonably stable? But might still be considered hacky... https://github.com/asmap/kartograf/blob/master/kartograf/util.py#L54

> annoying little period on test output is messing with the alignment.

Will test later to see if I know how to fix that. Let me know if you want to make further changes based on the util function above.

jurraca commented:
> I implemented it like this for the log outputs we generate on the Version

nice thanks, I somehow missed that. I pulled out the local and "remote" version checks of that function so it's reusable in the environment test. 
With the Nix version check, we can use that as our "expected" version to run, and don't have to hard code a version everywhere.

Having both `check_compatibility` and the `EnvironmentTest` is good I think. I removed the test assertion from the rpki-version check since it can fail in ways that alert the user. But the other version checks are strict failures.

jurraca commented:
Changed the setup to use pytest [e181ca3](https://github.com/asmap/kartograf/pull/26/commits/e181ca3e4d1b08431947b27ef67fa4b2696479d4). The main changes worth looking at beyond "it prints a version check" are the changes to `kartograf/util.py` for getting local and expected rpki-client versions.

jurraca commented:
rebased

fjahr reviewed (commented):
I think you can squash the commits here, per our offline conversation. It's currently easier for me to look at the final version (files changed) rather than the commits one-by-one because stuff that is changed once get's changed again (moving to unittest and then to pytest for example). Separate commits can still make sense when that is not the case and you separate things by functionality implemented and can keep them self-contained. But just if you want to, at this size/complexity it's not necessary. For example (off the top of my head, just to demonstrate what I mean), this could be 1 commit pure refactor of util.py, 1 commit update dependencies, 1 commit implement check command.

jurraca commented on check:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
yea didn't think of windows, thanks. changed it to python.

jurraca commented on flake.nix:
```diff
@@ -41,15 +41,15 @@
           ps.beautifulsoup4
           ps.numpy
+          ps.pandas
+          ps.pytest
```
yep, thanks.

jurraca commented on flake.nix:
```diff
@@ -58,7 +58,7 @@
       # * A NixOS module
       devShell = pkgs.mkShell {
-        packages = kartografDeps;
+        packages = kartografDeps ++ [pkgs.python311Packages.autopep8];
```
was trying out formatters and forgot about this, sorry.

jurraca commented on tests/environment_test.py:
```diff
@@ -0,0 +1,66 @@
+def test_installed_packages():
+    python_executable_path = PurePath(sys.executable)
+    python_env_path = python_executable_path.parents[1].as_posix()
+    required_packages = {
```
yea good point, added parsing for requirements.txt.

jurraca commented on kartograf/util.py:
```diff
@@ -51,37 +51,58 @@ def rir_from_str(maybe_rir):
+    rpki_path = subprocess.check_output(
+        ["which", "rpki-client"]).decode().strip()
+    try:
+        derivation = subprocess.check_output(["nix", "derivation", "show", rpki_path])
```
yea thanks, I was thinking of Nix as our versions "source of truth" in order to avoid having to set the RPKI version as a global var, but reverted to that.

jurraca commented:
The commits still aren't logical but it should be easier to read. lmk if you want it cleaner -- i'll approach this differently next time.

This uses the local (in the path) rpki-client version to `check_compatibility`, and uses that same function in the check as is used in `run`. Getting pytest added to the flake correctly resulted in some duplication but I couldn't figure out the right Nix incantation to compose python packages.

fjahr commented on kartograf/util.py:
```diff
@@ -4,6 +4,9 @@
 import time
+import json
+
+RPKI_VERSION = 9.3
```
nit: Could probably be moved into some kind of globals file as a follow-up

fjahr commented on check:
```diff
@@ -0,0 +1 @@
+python -m pytest -sq tests/environment_test.py
```
Hm, my comment about pytest being a dev dependency typically was also meant as removing it from check the check command. I think I wasn't very clear in my explanation. My point was that any testing library is usually just used for testing and not for actually running the software itself. Because of that it's usually a dev dependency because the people running the software itself don't need it. I would say the check command is something that is not for devs but rather for normal users. So it just gave me a bit of an ick that normal users would need to install a testing library to run it. Especially since the check command isn't even the actual software itself. But, as I mentioned, it's not a big deal. But with the current version it's kind of the worst of both worlds because it's now listed as a dev dependency but in order to run the check command it is still needed. So if a normal user would want to run the check command they would need to install the dev dependencies. I'm fine with each approach (removing pytest from check as a dependency or just making pytest not a dev dependency anymore) but it should be made consistent again. Sorry that my previous comment was confusing, I hope this makes sense.

jurraca replied:
Ok, yea I see, sorry I misunderstood the problem here. If we make this a script and not a test it makes more sense, see [014b99f](https://github.com/asmap/kartograf/pull/26/commits/014b99f574f8c7629457079e24dbdcb76d5cc6aa). This way the user only requires python, and everything else gets pretty printed whether its in the path or not. lmk if this works.

fjahr reviewed (commented):
Tested and it works. The output looks good for when everything is green but when I remove a python dep it looks kind of scary.  Is there a simple way to not print all the stuff below and just tell the users "you are missing X, please install it"?

```
$ ./check
Using rpki-client version 9.3 (recommended).
.Python version:           ✅ OK  (3.12)
.beautifulsoup4 version:   ✅ OK  (4.11.1)
numpy version:            ✅ OK  (1.26.4)
pandas version:           ✅ OK  (1.5.3)
requests version:         ✅ OK  (2.32.3)
tqdm version:             ✅ OK  (4.66.3)
F
========================================================================================================================== FAILURES ===========================================================================================================================
___________________________________________________________________________________________________________________ test_installed_packages ___________________________________________________________________________________________________________________

cls = <class 'importlib.metadata.Distribution'>, name = 'pandarallel'

    @classmethod
    def from_name(cls, name: str):
        """Return the Distribution for the given package name.

        :param name: The name of the distribution package to search for.
        :return: The Distribution instance (or subclass thereof) for the named
            package, if found.
        :raises PackageNotFoundError: When the named package's distribution
            metadata cannot be found.
        :raises ValueError: When an invalid value is supplied for name.
        """
        if not name:
            raise ValueError("A distribution name is required.")
        try:
>           return next(cls.discover(name=name))
E           StopIteration

/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/metadata/__init__.py:397: StopIteration

During handling of the above exception, another exception occurred:

    def test_installed_packages():
        python_executable_path = PurePath(sys.executable)
        python_env_path = python_executable_path.parents[1].as_posix()
        required_packages = __read_requirements_txt()
        for package, min_version in required_packages.items():
>           dist = distribution(package)

tests/environment_test.py:45:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/metadata/__init__.py:862: in distribution
    return Distribution.from_name(distribution_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'importlib.metadata.Distribution'>, name = 'pandarallel'

    @classmethod
    def from_name(cls, name: str):
        """Return the Distribution for the given package name.

        :param name: The name of the distribution package to search for.
        :return: The Distribution instance (or subclass thereof) for the named
            package, if found.
        :raises PackageNotFoundError: When the named package's distribution
            metadata cannot be found.
        :raises ValueError: When an invalid value is supplied for name.
        """
        if not name:
            raise ValueError("A distribution name is required.")
        try:
            return next(cls.discover(name=name))
        except StopIteration:
>           raise PackageNotFoundError(name)
E           importlib.metadata.PackageNotFoundError: No package metadata was found for pandarallel

/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/metadata/__init__.py:399: PackageNotFoundError
=================================================================================================================== short test summary info ===================================================================================================================
FAILED tests/environment_test.py::test_installed_packages - importlib.metadata.PackageNotFoundError: No package metadata was found for pandarallel
1 failed, 2 passed in 0.05s
```

jurraca commented:
thanks for catching that, fixed it and made the rpki-client version check raise or return appropriately too. 
User would either run `python check.py` or `python -m check` at the root of the project and everything should get pretty printed accordingly.

fjahr commented:
> If this is better, I'll squash and maybe add a note in the README about it.

Sounds good to me! Looks like there is also a merge conflict.

fjahr commented:
ACK, very nice!
